Make CBlock a vector of shared_ptr of CTransactions

sipa opened the pull request:
This is another preparation for #8580, avoiding performance problems and large code changes ahead of time (I was originally planning to do these afterwards, but that would temporarily introduce extra complexity).

In many places we've started using `std::shared_ptr<const CTransaction>` instead of `CTransaction` itself, as transactions are shared between many data structures (orphan pool, mempool, relay pool, partially-reconstructed transactions, ...), but a notable exception was `CBlock`, which so far always required a full copy in and out. If CTransaction is to become immutable, that will make things even worse when modifying the coinbase (https://github.com/bitcoin/bitcoin/pull/8580#issuecomment-243052533).

Advantages:
* The mining code no longer needs to create a deep copy of all transactions; the resulting CBlockTemplate just shares transactions with the mempool they're taken from.
* Compact block reconstruction no longer copies transactions (not from the prefilled structure to the internal partial block, and not from the partial block to the full block).
* Block processing no longer copies full transactions to txChanged.
* CTransaction becomes constructable-from-serialization, preparing for #8580.

Disadvantages:
* An extra pointer indirection and atomic increments/decrements in some places.

sipa committed:
Add serialization for unique_ptr and shared_ptr

sipa committed:
Add deserializing constructors to CTransaction and CMutableTransaction

sipa committed:
Make CBlock::vtx a vector of shared_ptr<CTransaction>

sipa committed:
Introduce convenience type CTransactionRef

theuni commented:
Concept ACK. _Very_ curious to see some numbers :)

sipa commented:
I created a branch with some follow-up *shared_ptr all the things* and other optimizations: https://github.com/sipa/bitcoin/commits/sharedblock2

gmaxwell commented:
tested ACK.

gmaxwell commented:
FWIW, I attempted to test the performance impact of this but it is a bit difficulty to do a fair comparison. I am reasonably confident that it does not produce a significant slowdown. (logically it should be a speedup, but I don't think my measurements were fair/reliable enough to determine if there was on or not).

sdaftuar commented:
I did a quick performance comparison running on 2 days of historical data, and though I was unable to quantify any meaningful performance difference, I agree that this doesn't appear to cause a significant slowdown.

sipa commented:
Benchmark results (25 reindex-chainstates, default dbcache, otherwise unloaded 24-core machine, measured by -debug=bench) on both master and #9125+#8580+#8589+a few shared_ptr optimizations I haven't PR'ed yet:
* Average: master 1490s, shared_ptr 1442s

Best master run:
```
2016-11-16 12:27:14   - Load block from disk: 1.25ms [273.57s]
2016-11-16 12:27:14     - Sanity checks: 0.14ms [70.95s]
2016-11-16 12:27:14     - Fork checks: 0.02ms [80.76s]
2016-11-16 12:27:14       - Connect 64 transactions: 11.06ms (0.173ms/tx, 0.020ms/txin) [711.67s]
2016-11-16 12:27:14     - Verify 550 txins: 11.09ms (0.020ms/txin) [718.24s]
2016-11-16 12:27:14     - Index writing: 0.03ms [5.25s]
2016-11-16 12:27:14     - Callbacks: 0.01ms [3.75s]
2016-11-16 12:27:14   - Connect total: 11.34ms [896.52s]
2016-11-16 12:27:14   - Flush: 0.26ms [86.70s]
2016-11-16 12:27:14   - Writing chainstate: 0.02ms [112.45s]
2016-11-16 12:27:14   - Connect postprocess: 0.42ms [86.88s]
2016-11-16 12:27:14 - Connect block: 13.30ms [1456.12s]
```

Best shared_ptr run:
```
2016-11-16 10:17:40   - Load block from disk: 1.61ms [283.93s]
2016-11-16 10:17:40     - Sanity checks: 0.18ms [76.82s]
2016-11-16 10:17:40     - Fork checks: 0.03ms [83.22s]
2016-11-16 10:17:40       - Connect 64 transactions: 17.43ms (0.272ms/tx, 0.032ms/txin) [717.76s]
2016-11-16 10:17:40     - Verify 550 txins: 17.47ms (0.032ms/txin) [724.51s]
2016-11-16 10:17:40     - Index writing: 0.02ms [5.34s]
2016-11-16 10:17:40     - Callbacks: 0.02ms [3.85s]
2016-11-16 10:17:40   - Connect total: 17.81ms [911.36s]
2016-11-16 10:17:40   - Flush: 0.40ms [87.91s]
2016-11-16 10:17:40   - Writing chainstate: 0.03ms [112.31s]
2016-11-16 10:17:40   - Connect postprocess: 0.08ms [25.61s]
2016-11-16 10:17:40 - Connect block: 19.94ms [1421.12s]
```

theuni commented on src/txmempool.cpp:
```diff
@@ -596,29 +596,29 @@ void CTxMemPool::removeConflicts(const CTransaction &tx, std::vector<std::shared
     }
-    BOOST_FOREACH(const CTransaction& tx, vtx)
+    for (const auto& tx : vtx)
     {
```
nit: could cache the dereferenced tx since it's used a few times

sipa replied:
Likewise.

theuni reviewed (approved):
utACK. Several nits, but nothing worth blocking.

theuni commented on src/main.cpp:
```diff
@@ -1682,9 +1682,9 @@ bool GetTransaction(const uint256 &hash, CTransaction &txOut, const Consensus::P
-            BOOST_FOREACH(const CTransaction &tx, block.vtx) {
-                if (tx.GetHash() == hash) {
-                    txOut = tx;
+            for (const auto& tx : block.vtx) {
```
Heh, https://github.com/bitcoin/bitcoin/pull/8580/commits/9b27eb1db1e4d1433f016dad381b52e8d986b685. Nevermind.

sipa commented on src/serialize.h:
```diff
@@ -25,6 +26,20 @@
+ * deserializing it from s. If T contains const fields, this
+ * is likely the only way to do so.
+ */
+struct deserialize_t {};
```
I'm not sure what you are suggesting here.

jtimon commented on src/primitives/transaction.h:
```diff
@@ -460,12 +464,25 @@ struct CMutableTransaction
+static inline CTransactionRef MakeTransactionRef() { return std::make_shared<const CTransaction>(); }
+static inline CTransactionRef MakeTransactionRef(const CTransaction& txIn) { return std::make_shared<const CTransaction>(txIn); }
+static inline CTransactionRef MakeTransactionRef(const CMutableTransaction& txIn) { return std::make_shared<const CTransaction>(txIn); }
+static inline CTransactionRef MakeTransactionRef(CMutableTransaction&& txIn) { return std::make_shared<const CTransaction>(std::move(txIn)); }
```
Where is this version used? I see calls that call std::move before calling that maybe should be using this instead?

In any case, it doesn't feel right to have so many versions of the function somehow. Specially one some with & and others with &&

theuni replied:
I believe the overloads could be replaced with:
```c++
template <typename T>
static inline CTransactionRef MakeTransactionRef(T&& txIn)
{
  return std::make_shared<const CTransaction>(std::forward<T>(txIn));
}
```

TheBlueMatt replied:
Might still be worth replacing the 6 functions with 3 using @theuni's suggestion

sipa commented:
@jtimon I believe all of them are used in successor PRs.

@theuni That overload won't work for the arguments that take a
CTransactionRef.

sipa commented:
@jtimon std::move just casts to an T&& argument. It's to indicate that the
called function may destroy the object.

jtimon commented:
utACK 9bc6cbc besides some nits from @theuni

theuni commented:
@sipa I'm unsure why you'd want to do that? Why wouldn't you just
```
CTransactionRef a(MakeTransactionRef());
CTransactionRef b(a);
CTransactionRef c(std::move(b));
```
?

sipa commented:
@theuni In #8580 I add a templated constructor to CWalletTx and CMerkleTx that just passes its arguments to MakeTransactionRef, avoiding duplicating various kinds of constructors in both.

sipa commented on src/main.cpp:
```diff
@@ -2884,7 +2885,7 @@ bool static ConnectTip(CValidationState& state, const CChainParams& chainparams,
     UpdateTip(pindexNew, chainparams);
 
-    for(unsigned int i=0; i < pblock->vtx.size(); i++)
+    for (unsigned int i=0; i < pblock->vtx.size(); i++)
```
You take one down, pass it around. Infinity cases of nits in the code.

sipa commented on src/main.cpp:
```diff
@@ -3109,13 +3110,13 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
         // ... and about transactions that got confirmed:
-        for(unsigned int i = 0; i < txChanged.size(); i++)
-            GetMainSignals().SyncTransaction(std::get<0>(txChanged[i]), std::get<1>(txChanged[i]), std::get<2>(txChanged[i]));
+        for (unsigned int i = 0; i < txChanged.size(); i++)
```
Not going to touch that now, as I believe #9014 will rewrite this anyway.

sipa commented on src/core_memusage.h:
```diff
@@ -69,8 +69,8 @@ static inline size_t RecursiveDynamicUsage(const CMutableTransaction& tx) {
-    for (std::vector<CTransaction>::const_iterator it = block.vtx.begin(); it != block.vtx.end(); it++) {
-        mem += RecursiveDynamicUsage(*it);
+    for (const auto& tx : block.vtx) {
+        mem += RecursiveDynamicUsage(*tx);
```
Fixed.

sipa commented:
Rebased, and addressed two nits by @Theuni:
* Added the memory usage for the control block of the vtx entries in CBlock.
* Changed deserialize_t to deserialize_type to avoid colliding with the reserved _t suffix.

sipa commented:
Some more benchmarks (fastest of 28 reindexes each, 300 MB dbcache, reindex chainstate up to block 295000):
* master: 1462.4s
* this PR: 1413.6s
* #8580: 1412.2s
* #8589: 1420.5s
* https://github.com/sipa/bitcoin/commits/sharedblock2: 1410.3s

TheBlueMatt commented on src/serialize.h:
```diff
@@ -25,6 +26,20 @@
+ *
+ * By convention, a constructor of a type T with signature
+ *
+ *   template <typename Stream> T::T(deserialize_t, Stream& s)
```
nit: comment out of date

TheBlueMatt reviewed (commented):
utACK 39e832c90012db03822f61009060712a47f7f81d...comments all dont matter so much

sipa commented:
Addressed all of @TheBlueMatt's nits.

laanwj commented:
Tested ACK b4e4ba4

arielgabizon commented on src/primitives/transaction.h:
```diff
@@ -460,6 +464,11 @@ struct CMutableTransaction
     }
 
+    template <typename Stream>
+    CMutableTransaction(deserialize_type, Stream& s) {
```
Right now there's some code redundancy - the only deserialize_type is deserialize,
so it's not used like here.
But I guess people think there may be in the future.
